lib/segment/src/index/field_index/geo_hash.rs

```diff
@@ -259,11 +244,57 @@ pub fn rectangle_hashes(rectangle: &GeoBoundingBox, max_regions: usize) -> Vec<G
         .expect("geo-hash coverage is empty")
 }
 
+/// Return as-high-as-possible with maximum of `max_regions`
+/// number of geo-hash guaranteed to contain a boundary defined by closed LineString.
+fn boundary_hashes(boundary: LineString, max_regions: usize) -> Vec<GeoHash> {
+    let geo_bounding_box = minimum_bounding_rectangle_for_boundary(&boundary);
+    let full_geohash_bounding_box: GeohashBoundingBox = geo_bounding_box.into();
+    let polygon = Polygon::new(boundary, vec![]);
+
+    (0..=GEOHASH_MAX_LENGTH)
+        .map(|precision| {
+            full_geohash_bounding_box
+                .geohash_regions(precision, max_regions)
+                .map(|hashes| {
+                    hashes
+                        .into_iter()
+                        .filter(|hash| check_polygon_intersection(hash, &polygon))
+                        .collect_vec()
+                })
+        })
+        .take_while(|hashes| hashes.is_some())
+        .last()
+        .expect("no hash coverage for any precision")
```

Comment:
This is a pattern prevalent in this file, as discussed we should remove them all at once in a follow up PR.

Tracked here https://github.com/qdrant/qdrant/issues/2645

Reply:
ack.

Reply:
The error handling issue has been started by an external contributor, you could give it a short review :)
https://github.com/qdrant/qdrant/pull/2701

Reply:
The fix was merged, you can rebase and use the new error handling :)

Reply:
merged :)